File: dp02_grid_traveler.go
package main

import "fmt"

func main() {
	// fmt.Println(gridTraveler(1, 1))
	// fmt.Println(gridTraveler(0, 1))
	// fmt.Println(gridTraveler(1, 0))
	// fmt.Println(gridTraveler(0, 0))
	// fmt.Println(gridTraveler(2, 3))
	// fmt.Println(gridTraveler(5, 3))
	// fmt.Println(gridTraveler(4, 4))

	memo := map[string]int{}

	fmt.Println(gridTravelerDy(1, 1, memo))
	fmt.Println(gridTravelerDy(0, 1, memo))
	fmt.Println(gridTravelerDy(1, 0, memo))
	fmt.Println(gridTravelerDy(0, 0, memo))
	fmt.Println(gridTravelerDy(2, 3, memo))
	fmt.Println(gridTravelerDy(5, 3, memo))
	fmt.Println(gridTravelerDy(10, 50, memo))
}

func gridTraveler(m, n int) int {
	if m == 0 || n == 0 {
		return 0
	}

	if n == 1 && n == 1 {
		return 1
	}

	return gridTraveler(m-1, n) + gridTraveler(m, n-1)
}

func gridTravelerDy(m, n int, memo map[string]int) int {
	if m == 0 || n == 0 {
		return 0
	}

	if n == 1 && n == 1 {
		return 1
	}

	var key string

	if n > m {
		key = fmt.Sprint("%d %d", m, n)
	} else {
		key = fmt.Sprint("%d %d", m, n)
	}

	if val, ok := memo[key]; ok {
		return val
	}

	memo[key] = gridTravelerDy(m-1, n, memo) + gridTravelerDy(m, n-1, memo)

	return memo[key]
}

File: dp02_grid_traveler.py
def grid_traveler_dy(m, n, memo={}):
    if not (m and n):
        return 0

    if n == m == 1:
        return 1
    if m > n:
        key = f"{m} {n}"
    else:
        key = f"{n} {m}"

    if key in memo:
        return memo[key]
    
    memo[key] = grid_traveler_dy(m-1, n, memo) + grid_traveler_dy(m, n-1, memo)

    return  memo[key]
    
def main():
    # print(grid_traveler(1, 1))
    # print(grid_traveler(1, 0))
    # print(grid_traveler(0, 0))
    # print(grid_traveler(3, 2))
    # print(grid_traveler(13, 18))

    print(grid_traveler_dy(1, 1))
    print(grid_traveler_dy(4, 4))
    print(grid_traveler_dy(10, 10))
    print(grid_traveler_dy(15, 15))

main()